### ISSUE TITLE
Disambiguated Properties Are Not Renamed

### ISSUE THREAD

**Issue by ChadKillingsworth** [PolymerLabs/PolymerRenamer#5]
The renamer doesn't appear to support the type-based property renaming of the compiler.

Rename map entries such as:

```
MyCustomElement_prototype$methodName:Ub
```

Seem to be ignored.

**Comment by ChadKillingsworth** [comment]
Actually now that I look at that, it's not a disambiguated property - it's a devirtualized prototype method. That's a bug in the compiler - not here.

**Comment by ChadKillingsworth** [comment]
Nevermind - I was right the first time. Repo case:

```
Polymer({
    is: "foo-bar",
    uncalledMethod: function(bar) {
        return bar.toString();
    }
});

Polymer({
    is: "foo-bar2",
    uncalledMethod: function(bar) {
        return bar.toString();
    }
});
```

With the `--use_types_for_optimization=true` flag, the renaming report is:

```
FooBar2Element_prototype$uncalledMethod:a
FooBarElement_prototype$uncalledMethod:b
```

With the `--use_types_for_optimization=false` flag, the renaming report is:

```
uncalledMethod:a
```

**Comment by robliao** [comment]
Yes, ambiguated and disambiguated properties optimizations are currently unsupported in the PolymerRenamer. The last mile of disambiguated properties is getting the transitively renamed symbol from JSCompiler to the map.

Using the disambiguated example above, there is an implicit rename of

```
uncalledMethod:FooBar2Element_prototype$uncalledMethod
uncalledMethod:FooBarElement_prototype$uncalledMethod
```

Reliably getting the type information to conclude that the symbol renamed in Polymer is safe is currently an unsolved problem. Not only do the definitions need to be renamed, but the expressions that reference the definitions also need to be renamed. This means the PolymerRenamer needs to know what is `FooBar` and `FooBar2`.

In other words, this is the tricky bit:

```
Rename Map:
foo$disambiguatedProperty:a
foo2$disambiguatedProperty:b

<template>
  <div>{{baz.disambiguatedProperty}}</div>
</template>
```

It isn't clear from this context whether or not `baz` is a `foo` or `foo2`.

The workaround at the moment is to avoid using `--compilation_level=ADVANCED_OPTIMIZATIONS`.

**Comment by ChadKillingsworth** [comment]
So a solution here may be to allow JSDoc type annotations in polymer templates. While not perfect, at least that gives responsibility to the user. Since the renamer users the parser from the compiler, JSDoc annotation parsing should be possible as well.

```
Rename Map:
foo$disambiguatedProperty:a
foo2$disambiguatedProperty:b

<template>
  <div>{{/** @type {foo2} */ (baz).disambiguatedProperty}}</div>
</template>
```

A shorter term solution might be to support disambiguated property renames on direct polymer element properties (not sub-paths). That would allow computed bindings as a workaround.

```
Rename Map:
foo$disambiguatedProperty:a
foo2$disambiguatedProperty:b

<dom-module id="foo-bar"><!-- use FooBarElement_prototype$ as a renaming prefix lookup -->
  <template>
    <div>{{getDisambiguatedProperty(baz)}}</div><!-- returns baz.disambiguatedProperty -->
  </template>
</dom>
```

`getDisambiguatedProperty` may itself be disambiguated, but that should be pretty easy to solve for since it will be prefixed by the id attribute of the parent `dom-module`.

**Comment by ChadKillingsworth** [comment]
We could also modify the property renaming report of the compiler to output property types - not just names. That would allow chained lookups in the map - which would be another part of the puzzle.

```
foo$disambiguatedProperty:a:bar
foo2$disambiguatedProperty:b:baz
bar$disambiguatedProperty:c:foobar
```
